lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -564,35 +532,53 @@ extendAddressPool !ix !pool
   where
     edge = Map.size (indexedKeys pool)
     isOnEdge = edge - fromEnum ix <= fromEnum (gap pool)
-    next = if ix == maxBound then mempty else nextAddresses @n
-        (accountPubKey pool)
+    next = if ix == maxBound then mempty else nextAddresses @n @c
+        (context pool)
         (gap pool)
-        (role @c)
         (succ ix)
 
 -- | Compute the pool extension from a starting index
 nextAddresses
-    :: forall (n :: NetworkDiscriminant) k.
+    :: forall (n :: NetworkDiscriminant) (c :: Role) k.
         ( MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+        , MkKeyFingerprint k Address
         , SoftDerivation k
+        , Typeable c
+        , Typeable n
         )
-    => k 'AccountK XPub
+    => ParentContext c k
     -> AddressPoolGap
-    -> Role
     -> Index 'Soft 'AddressK
     -> Map (KeyFingerprint "payment" k) (Index 'Soft 'AddressK, AddressState)
-nextAddresses !key (AddressPoolGap !g) !cc !fromIx =
+nextAddresses !ctx (AddressPoolGap !g) !fromIx =
     [fromIx .. min maxBound toIx]
-        & map (\ix -> (newPaymentKey ix, (ix, Unused)))
+        & map (\ix -> (mkPaymentKey ix ctx, (ix, Unused)))
         & Map.fromList
   where
     toIx = invariant
         "nextAddresses: toIx should be greater than fromIx"
         (toEnum $ fromEnum fromIx + fromEnum g - 1)
         (>= fromIx)
-    newPaymentKey = unsafePaymentKeyFingerprint @k
-        . (Proxy @n,)
-        . deriveAddressPublicKey key cc
+
+    mkPaymentKey ix = \case
+        ParentContextUtxoExternal acct ->
+            mkPaymentKeyFromAccXPub acct
+        ParentContextUtxoInternal acct ->
+            mkPaymentKeyFromAccXPub acct
+        ParentContextMutableAccount acct ->
+            mkPaymentKeyFromAccXPub acct  -- think about whether it makes sense
+        ParentContextMultisigScript _ payment delegation ->
+            mkPaymentKeyFromTemplates payment delegation
```

Comment:
Now this seems suspicious to me. when I saw that the `ParentContextMultisigScript` definition above included the parent key, I thought the parent key had been factored out up there. But it isn't use here to construct payment keys, which suggests that `payment` and `delegation` both contains the parent key. Thus, there are two places which contain redundant information. I prefer the approach where the parent keys are held by each template; 

Unless I am missing something, there shouldn't be any parent account key in the `ParentContextMultisigScript` itself.

Reply:
if we want to have a way to say which verification key was used in discovered script address, then probably we will need to drag along accXPub. It is used in `isShared` to construct keyHash with this accXPub and used index